In [1]:
import numpy as np
from scipy.stats import chi2

## 📌 Question 1 : Chi-Square Goodness-of-Fit (Fair Die)

A die is rolled 120 times. Observed counts per face:

| Face | 1 | 2 | 3 | 4 | 5 | 6 |
|------|---|---|---|---|---|---|
| Observed | 18 | 22 | 17 | 25 | 20 | 18 |

Test at α = 0.05 whether the die is fair using chi-square goodness-of-fit. Compute χ², df, and the critical value.

*If the die is fair every face has equal probability 1/6, so under H₀ each face should appear n/6 = 20 times. The chi-square statistic measures how far the observed counts stray from those equal expected counts. A large χ² means the die is likely biased; a small one means the deviations are just chance. Since we are comparing one set of observed frequencies against a fully specified theoretical distribution (no parameters estimated from the data), df = k - 1 = 5.*

#### Step 1 : Hypotheses

$$H_0: p_1 = p_2 = p_3 = p_4 = p_5 = p_6 = \frac{1}{6} \quad \text{(die is fair)}$$
$$H_1: \text{at least one } p_i \neq \frac{1}{6} \quad \text{(die is not fair)}$$

#### Step 2 : Why This Test

We have one categorical variable (die face) with $k = 6$ categories and a fixed theoretical distribution under $H_0$. The chi-square goodness-of-fit test is designed exactly for this: it compares a single sample of observed frequencies against expected frequencies derived from the null distribution. All expected counts = 20 ≥ 5, so the large-sample approximation is valid.

#### Step 3 : Test Statistic

$$\chi^2 = \sum_{i=1}^{k} \frac{(O_i - E_i)^2}{E_i}, \quad df = k - 1$$

where $O_i$ are observed counts and $E_i = n/k$ are expected counts under a fair die.

In [2]:
observed = np.array([18, 22, 17, 25, 20, 18])   # observed face counts
n        = observed.sum()                          # total rolls = 120
alpha    = 0.05                                    # significance level
k        = len(observed)                           # number of categories

In [3]:
expected  = np.array([n / k] * k)               # expected counts under fair die
chi_stat  = sum((observed - expected)**2 / expected)

print(f'Expected counts : {expected}')
print(f'chi_stat        : {chi_stat:.4f}')

Expected counts : [20. 20. 20. 20. 20. 20.]
chi_stat        : 2.3000


#### Step 4 : Degrees of Freedom

$$df = k - 1 = 6 - 1 = 5$$

No parameters are estimated from the data (the fair-die probabilities are fully specified), so we lose only 1 degree of freedom for the total constraint $\sum O_i = n$.

In [4]:
df = k - 1      # degrees of freedom = 5
print(f'df : {df}')

df : 5


#### Step 5 : Critical Value Method

In [5]:
chi_crit = chi2.ppf(1 - alpha, df)     # one-tailed (right tail) critical value
print(f'chi_stat : {chi_stat:.4f}')
print(f'chi_crit : {chi_crit:.4f}')

chi_stat : 2.3000
chi_crit : 11.0705


**chi_stat (2.30) < chi_crit (11.071) → Fail to Reject H₀**

> Note: chi-square goodness-of-fit is always a **right-tailed** test — only an unusually *large* χ² is evidence against H₀. The critical value uses `chi2.ppf(1 - alpha, df)`, not `1 - alpha/2`.

#### Step 6 : p-value Method

In [6]:
pval = chi2.sf(chi_stat, df)            # right-tail p-value
print(f'p-val : {pval:.4f}')

p-val : 0.8063


**p-val (0.806) > 0.05 → Fail to Reject H₀**

#### ✅ Final Conclusion

| Metric | Value |
|---|---|
| Total rolls $n$ | 120 |
| Categories $k$ | 6 |
| Expected per face | 20 |
| $\chi^2$ statistic | 2.30 |
| Degrees of freedom | 5 |
| $\chi^2$ critical (right-tail, α = 0.05) | 11.071 |
| p-value | 0.806 |
| Decision | **Fail to Reject H₀** |

There is insufficient evidence to conclude the die is unfair. The observed counts (18, 22, 17, 25, 20, 18) are consistent with random variation expected from a fair die (χ² = 2.30, p = 0.806).